### ISSUE TITLE
NullReferenceException on code with dynamic casts

### ISSUE THREAD

**Issue by fvilers** [Particular/Particular.CodeRules#4]
Hi,

The analyzer throws an exception on code with dynamic casts.

For example, this code:
```
await ((dynamic)handler).HandleAsync((dynamic)command);
```
creates a warning compilation:
```
Warning	AD0001	Analyzer 'Particular.CodeRules.ConfigureAwait.ConfigureAwaitAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.
```
Best regards,
Fabian

**Comment by WilliamBZA** [comment]
Thanks @fvilers, that's quite a catch.

I'll dig around and see if we can sort that out for you.

**Comment by fvilers** [comment]
You're welcome. The second dynamic (`.HandleAsync((dynamic)command);`) is unecessary. The first one is probably causing the issue.

**Comment by WilliamBZA** [comment]
This is occurring because the `ITypeSymbol` for `dynamic` doesn't have a Containing Namespace that we can inspect.

We have 2 options:

1. Assume that anything within an `await` expression should have a `ConfigureAwait(...)`
2. Ignore the error and don't run the analyzer for `dynamic` types

Now while it's unlikely that something other than a `Task` will be awaited, it is possible. So I'm a little hesitant to make the assumption required by `1`.  I think it's probably better to go for option `2` meaning that we don't show analyzer warnings for `dynamic` types.

What do you think @distantcam @DavidBoike @fvilers?

**Comment by fvilers** [comment]
Regarding your comment, I'd agree with `2`. Using dynamic means you must know in advance what methods are available and that you pay extra attention on this part of the code. So if you do, you have to add the right `.ConfigureAwait()` by yourself.

Is it feasible that the code rule raises a warning instead of an error for this specific issue?

**Comment by DavidBoike** [comment]
@WilliamBZA agree with Option 2. If you go dynamic you are by definition on your own. It's enough of an edge case that it doesn't merit moving mountains (and burning cycles) attempting to detect.

Although, why does the [code in question](https://github.com/Particular/Particular.CodeRules/blob/master/src/Particular.CodeRules/ConfigureAwait/ConfigureAwaitAnalyzer.cs#L44) need to inspect the namespace and type name separately? On ITypeSymbol, should be able to just use `type.ToString() == "System.Threading.Tasks.Task` - what would happen in the dynamic case then?

A warning would just be an additional diagnostic with different settings, but then you'd essentially be guaranteeing a warning on any awaited dynamic. Is that worth it?

**Comment by distantcam** [comment]
Option 2. When in doubt don't cause a false negative, and don't crash. ;)

**Comment by timbussmann** [comment]
hey @fvilers 
I just wanted to let you know that we fixed this issue with version 0.2.1 which is now available on nuget. This version contains option 2 as described in the comments, which means it won't analyze dynamic return types for missing `ConfigureAwait`s.

Thanks again for raising this issue 👍

**Comment by fvilers** [comment]
Great news! Unfortunately, I do not work anymore on this project so I can't test this issue again. Good work!
